pkg/policy/mapstate.go

```diff
@@ -220,32 +244,176 @@ func getNets(identities Identities, ident uint32) []*net.IPNet {
 	return identities.GetNetsLocked(id)
 }
 
+// NewMapState creates a new MapState interface
+func NewMapState(initMap map[Key]MapStateEntry) MapState {
+	return newMapState(initMap)
+}
+
+func newMapState(initMap map[Key]MapStateEntry) *mapState {
+	m := &mapState{
+		allows: make(map[Key]MapStateEntry),
+		denies: make(map[Key]MapStateEntry),
+	}
+	for k, v := range initMap {
+		m.Insert(k, v)
+	}
+	return m
+}
+
+// Get the MapStateEntry that matches the Key.
+func (ms *mapState) Get(k Key) (MapStateEntry, bool) {
+	if ms == nil || (ms.allows == nil && ms.denies == nil) {
+		return MapStateEntry{}, false
+	}
+	if ms.denies == nil {
+		v, ok := ms.allows[k]
+		return v, ok
+	} else if ms.allows == nil {
+		v, ok := ms.denies[k]
+		return v, ok
+	} else {
+		v, ok := ms.allows[k]
+		if ok {
+			return v, ok
+		}
+		v, ok = ms.denies[k]
+		return v, ok
+	}
+}
+
+// Insert the Key and matcthing MapStateEntry into the
+// MapState
+func (ms *mapState) Insert(k Key, v MapStateEntry) {
+	isDeny := v.IsDeny
+	if ms != nil && ((isDeny && ms.denies != nil) || (!isDeny && ms.allows != nil)) {
+		if isDeny {
+			ms.denies[k] = v
+		} else {
+			ms.allows[k] = v
+		}
+	}
+}
+
+// Delete removes the Key an related MapStateEntry.
+func (ms *mapState) Delete(k Key) {
+	if ms != nil {
```

Comment:
Why do we check for nil here? Do we rely on being able to use `*mapState` with nil value somewhere?